### PR TITLE
Implement a hideItemPredicate on ORKFormItem to make each hideable

### PR DESCRIPTION
**Pull request by eschramm**
This adds the pieces needed for https://github.com/CareEvolution/CEVResearchKit/issues/134.

Here an `NSPredicate` can be set on an `ORKFormItem` which  if true will cause it to be hidden in an `ORKFormStepViewController`. Due to the differences between CEVRelease-1.x and the base ResearchKit 2.0, I needed to apply the changes manually in both versions. Note the following branches:

- [Schramm/hideFormItemWIthPredicate-1.x](https://github.com/CareEvolution/ResearchKit/tree/Schramm/hideFormItemWIthPredicate-1.x) - implementation for our local use based on [CEVRelease-1.x](https://github.com/CareEvolution/ResearchKit/tree/CEVRelease-1.x). This is what I'm requested to be merged in this PR.
- [Schramm/demoTestHideFormItemWIthPredicate-1.x](https://github.com/CareEvolution/ResearchKit/tree/Schramm/demoTestHideFormItemWIthPredicate-1.x) - a convenient branch with 5 cases built out in ORKCatalog to test/demo changes in [Schramm/hideFormItemWIthPredicate-1.x](https://github.com/CareEvolution/ResearchKit/tree/Schramm/hideFormItemWIthPredicate-1.x)
- [Schramm/hideFormItemWIthPredicate-2.0](https://github.com/CareEvolution/ResearchKit/tree/Schramm/hideFormItemWIthPredicate-2.0) - implementation to submit PR against [ResearchKit/ResearchKit/master](https://github.com/ResearchKit/ResearchKit/tree/master). I would appreciate feedback on this as well prior to submitting the PR.
- [Schramm/demoTestHideFormItemWIthPredicate-2.0](https://github.com/CareEvolution/ResearchKit/tree/Schramm/demoTestHideFormItemWIthPredicate-2.0) - a convenient branch with the same 5 cases built out in ORKCatalog to test/demo changes in [Schramm/hideFormItemWIthPredicate-2.0](https://github.com/CareEvolution/ResearchKit/tree/Schramm/hideFormItemWIthPredicate-2.0)

In a nutshell, I'm attempting to preserve the full unfiltered `ORKFormStep` with all the `ORKFormItems`, but modify the state of the tableview based on the hide predicates and respond to live changes.

### CODE REVIEW

**Comment by chrisnowak** [comment]
@eschramm couple questions...
- How does this work with required form items?  If a required form item is hidden - is it still required and therefore the next button will not be available to be hit unless the form item is shown?   (note - I think the correct behavior is that if a form item is hidden it should be treated as non-existent as far as "Required" status goes...)
- How does this work with regards to results...when you hit "next" does it still store results for the hidden form items.  My sense is that we would NOT want to store a result for hidden form items just like we would not want to store a result for skipped question steps

**Comment by eschramm** [comment]
@chrisnowak - thanks for those considerations. I have made adjustments to account for those cases as you suggest:

- `ORKFormItem`s that are filtered and required (not optional) become unrequited and do not prevent navigation/advancing to the next step.
- if an `ORKFormItem` is hidden via the `hideItemPredicate`, it is removed from the result entirely versus showing a (null) result. This maintains consistent behavior with navigator rule-skipped steps

File: ResearchKit/Common/ORKFormStep.h [hunk 1]
```diff
@@ -195,6 +195,11 @@ ORK_CLASS_AVAILABLE
  */
 @property (nonatomic, copy, readonly, nullable) ORKAnswerFormat *answerFormat;
 
+/**
+ A predicate that when true, hides the item from display.
+ */
+@property (nonatomic, nullable) NSPredicate *hideItemPredicate;
```
Comment (overcyn) [inline ResearchKit/Common/ORKFormStep.h]: Minor but the word `item` in `hideItemPredicate` seems redundant.

File: ResearchKit/Common/ORKFormStepViewController.m [hunk 2]
```diff
@@ -692,6 +746,10 @@ - (ORKStepResult *)result {
     
     NSMutableArray *qResults = [NSMutableArray new];
     for (ORKFormItem *item in items) {
+        
+        if ([_filteredForms containsObject:item]) {
```
Comment (overcyn) [inline ResearchKit/Common/ORKFormStepViewController.m]: Is it necessary to add an ivar for the filtered forms vs evaluating the predicate again?

Comment (eschramm) [reply ResearchKit/Common/ORKFormStepViewController.m]: I originally wanted to do it that way, but to evaluate the predicate, you need a current version of what this function outputs which then leads to a circular dependency :(

Comment (overcyn) [reply ResearchKit/Common/ORKFormStepViewController.m]: Interesting, if thats the case it seems like the circular dependency is still there and we are performing a single cycle every time `filterSections:` is called? Is there a way to get the full results with no filtering, and pass that into the predicate? Otherwise I feel like you could have ambiguous predicates where item A depends on item B, and item B depends on item A and it's unclear which takes precedent.

File: ResearchKit/Common/ORKFormStep.h (hunk 1 above)
Comment (eschramm) [reply ResearchKit/Common/ORKFormStep.h]: I agree - updated to `hidePredicate`

File: ResearchKit/Common/ORKFormStepViewController.m (hunk 2 above)
Comment (eschramm) [reply ResearchKit/Common/ORKFormStepViewController.m]: It is a tricky "chicken or egg" scenario. But I think they are not dependent upon each other because when `filterSections:` gets called, it's emptying out the `_filteredForms` array before requesting the current `result` so that logic technically won't rely on the predicate at that time. However, when navigating into the next step, result will be called via navigation methods (more specifically via `[ORKTaskViewController stepViewController:didFinishWithNavigationDirection:]`). In that case the `_filteredForms` will be populated with the latest state at completion of the FormStep and any filtered results will be removed before passing up the hierarchy. 
I still think there's a possibility to have interfering predicates like you suggest. Without have a validation system at construction, I'm not sure how to solve that problem.

Comment (overcyn) [reply ResearchKit/Common/ORKFormStepViewController.m]: Ah I see what you're saying. That makes sense. LGTM
